wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth/EmbeddedOAuthAPI.java

```diff
@@ -123,20 +125,25 @@ public Response callback(UriInfo uriInfo, @Nullable List<String> errorValues)
       LOG.error(e.getMessage(), e);
     }
     final String redirectAfterLogin = getParameter(params, "redirect_after_login");
-    return Response.temporaryRedirect(URI.create(redirectAfterLogin)).build();
+    URI uri;
+    try {
+      uri = URI.create(redirectAfterLogin);
+    } catch (IllegalArgumentException e) {
+      // the redirectUrl was decoded by the CSM provider, so we need to encode it back.
```

Comment:
It is not clear why IllegalArgumentException is the indicator of decoded url

Reply:
This exception is thrown when we try to create a new URL from a string with unsupported characters like `{` and `}`. Initially such symbols are encoded, but BitBucket Server decodes them so they become unsupported.